### PR TITLE
SC2: Content update

### PR DESCRIPTION
**Pull request by Ziktofel**
## What is this fixing or adding?
Feature highlights:
- Adds many content to the SC2 game
- Allows custom mission order
- Adds race-swapped missions for build missions (except Epilogue and NCO)
- Allows War Council Nerfs (Protoss units can get pre - War Council State, alternative units get another custom nerf to match the power level of base units)
- Revamps Predator's upgrade tree (never was considered strategically important)
- Adds some units and upgrades
- Locked and excluded items can specify quantity
- Key mode (if opt-in, missions require keys to be unlocked on top of their regular regular requirements
- Victory caches - Victory locations can grant multiple items to the multiworld instead of one 
- Fixes the following issues:
  - https://github.com/ArchipelagoMW/Archipelago/issues/3531 
  - https://github.com/ArchipelagoMW/Archipelago/issues/3548

## How was this tested?
- Test async game
- Verification run (over all the game's content) to spot the bugs, so all the checks work
- Unit tests

## If this makes graphical changes, please attach screenshots.
- Client got simplified, dropped unneeded tabs:
Snímek obrazovky pořízený 2025-08-11 20-33-07
- Tracker redesign (more dynamically generated with item descriptions as tooltips):
Snímek obrazovky pořízený 2025-08-11 20-34-07

### CODE REVIEW

**Comment by agilbert1412** [comment]
I was about to drop an approval, based on having played in the beta and having had close to no issues, so I trust this.
But then I saw the line count, and I figured a play-only review without code diving would probably not qualify as "approved"
So take this comment as an... endorsement? I guess

**Comment by Ziktofel** [comment]
I've reviewed the code myself from contributors over the whole dev cycle as it's long as there are many changes that must blend in together

**Comment by Exempt-Medic** [comment]
> I was about to drop an approval, based on having played in the beta and having had close to no issues, so I trust this. But then I saw the line count, and I figured a play-only review without code diving would probably not qualify as "approved" So take this comment as an... endorsement? I guess

It's definitely a **massive** PR that does all sorts of things (just look at the labels), but the endorsement means at least a little something, so thanks

File: WebHostLib/static/assets/sc2Tracker.js [hunk 1]
```diff
@@ -1,49 +1,31 @@
...
+        updateSection('section-zerg', fakeDOM);
+        updateSection('section-protoss', fakeDOM);
+        updateSection('section-nova', fakeDOM);
+        updateSection('section-kerrigan', fakeDOM);
+        updateSection('section-keys', fakeDOM);
+        updateSection('section-locations', fakeDOM);
+      };
+      ajax.open('GET', url);
+      ajax.send();
+    }, 15000)
```
Comment (Berserker66) [inline WebHostLib/static/assets/sc2Tracker.js]: I mentioned this on discord but it seems to have gone ignored; there's no point in updating every 15 seconds when the server  only  updates up to once per minute.
Since this tracker was last updated we now also have an API extension that tells you exactly which second each minute this occurs, so traffic can be cut into a fourth while also improving latency.

File: Starcraft2Client.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 import ModuleUpdate
 ModuleUpdate.update()
 
-from worlds.sc2.Client import launch
+from worlds.sc2.client import launch
```
Comment (Berserker66) [inline Starcraft2Client.py:6]: We're moving away from client .exe's entirely. The client should be moved into the Launcher.

File: worlds/LauncherComponents.py [hunk 3]
```diff
@@ -230,7 +230,7 @@ def install_apworld(apworld_path: str = "") -> None:
     # ChecksFinder
     Component('ChecksFinder Client', 'ChecksFinderClient'),
     # Starcraft 2
-    Component('Starcraft 2 Client', 'Starcraft2Client'),
+    Component('Starcraft 2 Client', 'Starcraft2Client', game_name='Starcraft 2', supports_uri=True),
```
Comment (Berserker66) [inline worlds/LauncherComponents.py]: This is a list of old components  that are not yet defined inside the world, so instead of updating the legacy listing it should be moved into the world.

File: WebHostLib/static/assets/sc2Tracker.js (hunk 1 above)
Comment (Ziktofel) [reply WebHostLib/static/assets/sc2Tracker.js]: Fixed in 201673ded

File: WebHostLib/templates/tracker__Starcraft2.html [hunk 4]
```diff
@@ -1,1092 +1,2235 @@
...
-            </tr>
-        </table>
+    <div id="filler-items" class="tracker-section">
+      <div class="section-title">
+        <input type="checkbox" class="collapse-section">
+        <h2>Filler Items</h2>
+      </div>
+      <div id="section-filler" class="section-body">
+        <div class="item-counter">
+          <img src="https://matthewmarinets.github.io/ap_sc2_icons/icons/blizzard/icon-mineral-nobg.png" title="Additional Starting Minerals&#10;Increases the starting minerals for all missions.">
```
Comment (Berserker66) [inline WebHostLib/templates/tracker__Starcraft2.html]: With just how many icons there are, it would be nice, albeit not required, to use an atlas, as it could reduce the initial chug of having to load hundreds of images down to about 1%.

**Comment by Berserker66** [comment]
I've looked at the stuff that touches core/webhost and left some comments regarding those things, haven't looked at the stuff inside the world's folder at all yet.

File: WebHostLib/templates/tracker__Starcraft2.html (hunk 4 above)
Comment (Ziktofel) [reply WebHostLib/templates/tracker__Starcraft2.html]: @MatthewMarinets Can you look into this? Also I want to know how difficult it'd be to change due to this is auto-generated

Comment (Ziktofel) [reply WebHostLib/templates/tracker__Starcraft2.html]: @Berserker66 Also doesn't HTTP/2 solve this so these workarounds aren't needed?

File: Starcraft2Client.py (hunk 2 above)
Comment (Ziktofel) [reply Starcraft2Client.py:6]: I'm keeping this for now for Python 3.13 (as the client works there and Launcher doesn't). I've added this as deprecated fro now

File: worlds/LauncherComponents.py (hunk 3 above)
Comment (Ziktofel) [reply worlds/LauncherComponents.py]: Moved the definition to the world itself

File: WebHostLib/templates/tracker__Starcraft2.html (hunk 4 above)
Comment (MatthewMarinets) [reply WebHostLib/templates/tracker__Starcraft2.html]: I likely won't be available to look into this for about a week; IRL events are afoot. There's a chance I find the time in the next couple days, but after that I'm away from home, so don't let this hold up the PR (I can make a follow-up PR later if it comes to that).

On the icon repo side, I'd be concerned about how to go about generating/versioning the atlas for next release, as the coordinates for each individual icon may shuffle around on regeneration and thus there would be a risk of going out-of-sync with the tracker. Shouldn't be a blocker for a release version that gets frozen, though. I could probably do a one-time operation with `magick` to combine all the images into one, push that, update the generator to point all img tags at the one atlas, and then generate an additional .css to store all the `clip-path` / `transform` attributes necessary to display correctly. I haven't worked much with atlases in css before, so if anyone reading this has tips / knows a better way to stretch than `transform`, that would be appreciated.

re: HTTP/2, I doubt that would fix anything. Fundamentally the issue is that there's something like 700 `` tags here, and each of those is fetched as an independent request. My understanding is browsers batch those requests and have a limit on how many pending requests are allowed at a time before the next batch can start; HTTP/2 or QUIC wouldn't change that. An atlas would mean loading just 1 image and modifying the elements to only look at a small window of it; so just one request.

File: worlds/LauncherComponents.py (hunk 3 above)
Comment (MatthewMarinets) [reply worlds/LauncherComponents.py]: The moved definition in `__init__.py` calls itself "Factorio Client", which is incorrect.

I think we should also probably double-check URIs still work. I originally tested by clicking links in a webhost a few months back, but I think I had a time where it didn't work some months after.

File: WebHostLib/templates/tracker__Starcraft2.html (hunk 4 above)
Comment (MatthewMarinets) [reply WebHostLib/templates/tracker__Starcraft2.html]: Spent more time than I probably should have on it this evening. Got it working on chromium-based browsers, the loading times are indeed much faster and everything pops in at the same time. See [the PR](https://github.com/Ziktofel/Archipelago/pull/508)

The bad news is that it totally breaks on Firefox. FF doesn't seem to have the necessary accuracy on the `clip-path` coordinates, so all the icon boundaries are just way off and they include chunks of other icons around them. Naively attempting to set the `round` parameter on the `clip-path` to the actual units I'm working with (some fraction of a percent) didn't help. I additionally had to add a whole bunch of extra nesting to the HTML to get the proper order of operations on `transform` and `filter`, which led to a lot of bugs I fixed and possibly more I haven't found yet.

If someone knows something obvious I'm doing wrong, let me know.

The only further approach I can think of is to tile the atlas in 2 dimensions rather than just one, and doing a bit more math to extract clip/scale parameters, which should get sqrt(793) ~= 28x more accuracy on the clipping. If that's not enough, I guess I'd have to break things up into multiple atlases? I'm not getting to that tonight, and honestly I'm feeling burned enough by _another_ thing breaking on firefox I'm not particularly inclined to continue with this project.

Comment (Ziktofel) [reply WebHostLib/templates/tracker__Starcraft2.html]: Patched for the Firefox by using `overflow: hidden` for the image's parent element. Everything works then

File: WebHostLib/static/assets/sc2Tracker.js [hunk 5]
```diff
@@ -1,49 +1,42 @@
...
+    document.getElementById(sectionName).innerHTML = fakeDOM.getElementById(sectionName).innerHTML;
+}
 
-      // Create a fake DOM using the returned HTML
-      const domParser = new DOMParser();
-      const fakeDOM = domParser.parseFromString(ajax.responseText, 'text/html');
+window.addEventListener('load', () => {
+    // Reload tracker every 15 seconds
+    const url = window.location;
+    const targetSecond = parseInt(document.getElementById('player-tracker').getAttribute('data-second')) + 2;
```
Comment (Berserker66) [inline WebHostLib/static/assets/sc2Tracker.js]: 1: there is no  indication sections of code are copy-pasted from common tracker
2: why 2 instead of 3 seconds?

Comment (Ziktofel) [reply WebHostLib/static/assets/sc2Tracker.js]: @MatthewMarinets Can you look into that?

Comment (MatthewMarinets) [reply WebHostLib/static/assets/sc2Tracker.js]: Should there be credit for a copy this small? I was understanding this was just the "recommended" way of exercising the API, but adding a note on where it came from isn't hard.

2s instead of 3s was just a "try it and see if it works" change; it worked, and is easy enough to change later. There's no comment attached to 3s saying why it's there, so I just assumed it was a safety margin.

I can make another PR to fix these things.

Comment (Ziktofel) [reply WebHostLib/static/assets/sc2Tracker.js]: Solved in https://github.com/Ziktofel/Archipelago/pull/509

Comment (Berserker66) [reply WebHostLib/static/assets/sc2Tracker.js]: This isn't about credit, but keeping track of code that is likely to see more copy pasting. We have several such cases in AP where someone wrote something, it was copied, sometimes modified and  sometimes later unified. Keeping track of the origin early avoids later detective work.

As for making the time uniform, again for similar reasons,  so it can be more easily unified later. But also if there is no point in the change then it shouldn't  be made, else the next person changes it to 1 then another to 4 and another makes a mistakes and it doesn't work and  in  the end we have 30 different trackers that all behave slightly differently and it'll be a clusterfuck to try to fix anything.

File: worlds/sc2/__init__.py [hunk 6]
```diff
@@ -52,90 +86,279 @@ class SC2World(World):
...
-    final_mission_id: int
-    victory_item: str
-    required_client_version = 0, 4, 5
+    item_name_groups = item_groups.item_name_groups  # type: ignore
+    location_name_groups = location_groups.get_location_groups()
+    locked_locations: List[str]
+    """Locations locked to contain specific items, such as victory events or forced resources"""
+    location_cache: List[Location]
+    final_missions: List[int]
+    required_client_version = 0, 6, 3
```
Comment (Berserker66) [inline worlds/sc2/__init__.py]: This should probably be 0.6.4 by now.

Comment (Ziktofel) [reply worlds/sc2/__init__.py]: Version bumped

File: worlds/sc2/client.py [hunk 7]
```diff
@@ -0,0 +1,2352 @@
...
+    async def server_auth(self, password_requested: bool = False) -> None:
+        self.game = STARCRAFT2
+        if password_requested and not self.password:
+            await super(SC2Context, self).server_auth(password_requested)
+        await self.get_username()
+        await self.send_connect()
+        if self.ui:
+            self.ui.first_check = True
+
+    def is_legacy_game(self):
```
Comment (Berserker66) [inline worlds/sc2/client.py:682]: Does it actually still support the WOL only campaign? Has anyone tested that?

**Review by Berserker66 — APPROVED** [review]
Read through the code, didn't find any new security issues. Played a couple rounds. Found a bug that was already fixed and a couple oddities that were explained. Sadly the existing security issue of very old protobuf is not fixed with this, so everyone, including me, who is subscribed to security notifications for the AP repo will continue to get spammed weekly.

My personal grievance that /received is overwritten also remains.

**Comment by Ziktofel** [comment]
I plan to address the protobuf issue in the next dev cycle. It's on the priority list. I plan the SC2's next dev cycle to be mainly maintenance in order to fix those issues

File: worlds/sc2/client.py (hunk 7 above)
Comment (Ziktofel) [reply worlds/sc2/client.py:682]: Yes, that's the support for games < 0.4.5 that got kept with the release. How's the support length schedule?
